### PR TITLE
Merge plugin manifest automatically when building individual plugins with the `builder` plugin

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
* Merge `plugin_manifest.yaml` automatically when building individual plugins with the `builder` plugin

* The `builder` plugin v1.0.0 or greater, supports generating the correct `plugin_manifest.yaml` and
`plugin_bundle.tar.gz` when the user runs `make plugin-build PLUGIN_NAME="<plugin-name>"` individually
multiple times to build different plugins possibly with different plugin versions. For example:

```sh
# Build `foo`. This will generate `plugin_bundle.tar.gz` containing just `foo:v1.0.0` plugin
make plugin-build PLUGIN_NAME="foo" PLUGIN_BUILD_VERSION=v1.0.0

# Build `bar`. This will generate `plugin_bundle.tar.gz` containing `foo:v1.0.0` and `bar:v2.0.0`
make plugin-build PLUGIN_NAME="bar" PLUGIN_BUILD_VERSION=v2.0.0

# Build `baz`. This will generate `plugin_bundle.tar.gz` containing `foo:v1.0.0`, `bar:v2.0.0`, `baz:v3.0.0`
make plugin-build PLUGIN_NAME="baz" PLUGIN_BUILD_VERSION=v3.0.0
```

**Note:** Because the builder plugin will automatically create merged plugin bundle if the `plugin_manifest.yaml` already exists,
To remove all the already built plugins and start from fresh, please configure environment variable `PLUGIN_BUNDLE_OVERWRITE=true`.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR
```
$ make prepare-builder
cd cmd/plugin/builder && go build -o /Users/anujc/Documents/code/tanzu-cli/bin/builder .
$ rm -rf artifacts/
$ $ git describe --tags 
v1.0.0-dev-37-g569df492
```

* Build just the `builder` plugin and notice the plugin manifest file contains only the `builder` plugin with the correct version
```
$ make plugin-build PLUGIN_NAME=builder
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-01T23:39:14.758537-07:00
plugins:
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v1.0.0-dev
```

* Now, build just `test` plugin and notice that `test` plugin gets added to the plugin_manifest file along with `builder`
```
$ make plugin-build PLUGIN_NAME=test
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-01T23:41:02.373134-07:00
plugins:
    - name: test
      target: global
      description: Test the CLI
      versions:
        - v1.0.0-dev
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v1.0.0-dev
```

* Now, force CLI to build plugins with version `v2.0.0` and build only `builder` plugin and notice that `builder` plugin version in plugin_manifest gets updated
```
$ export PLUGIN_BUILD_VERSION=v2.0.0
$ make plugin-build PLUGIN_NAME=builder
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-01T23:42:39.901108-07:00
plugins:
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v2.0.0
    - name: test
      target: global
      description: Test the CLI
      versions:
        - v1.0.0-dev
```

* While keeping the `PLUGIN_BUILD_VERSION=v2.0.0` run make a target to build all plugins and notice both the plugin entry is updated in plugin_manifest with version v2.0.0
```
$ make plugin-build
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-01T23:43:49.704985-07:00
plugins:
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v2.0.0
    - name: test
      target: global
      description: Test the CLI
      versions:
        - v2.0.0
```

* Run the `make plugin-build` again after upsetting the `PLUGIN_BUILD_VERSION` and verify plugin_manifest gets updated again with the right plugin versions.

```
# Unset PLUGIN_BUILD_VERSION so CLI uses the default git tag
$ make plugin-build
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-01T23:45:08.591372-07:00
plugins:
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v1.0.0-dev
    - name: test
      target: global
      description: Test the CLI
      versions:
        - v1.0.0-dev
```

* Use `PLUGIN_BUNDLE_OVERWRITE` to overwrite the plugin bundle and manifest file.
```
$ make plugin-build PLUGIN_NAME=builder PLUGIN_BUNDLE_OVERWRITE=true
.
.
$ cat artifacts/plugins/plugin_manifest.yaml
created: 2023-08-02T11:35:49.47775-07:00
plugins:
    - name: builder
      target: global
      description: Build Tanzu components
      versions:
        - v1.0.0-dev
```

### Release note

```release-note
Merge `plugin_manifest.yaml` automatically when building individual plugins with the `builder` plugin
```

### CODE REVIEW

**Comment by marckhouzam** [comment]
Thanks for doing this.
I'm trying to understand the developer workflow here.

The flow you are trying to address I assume is:
1. in a repo with plugins A and B
2. build plugin A only
3. then build plugin B only
4. then publish both plugins together

But what if the flow is:
1. in a repo with plugins A and B
2. build plugin A only
3. do development for a couple of weeks on plugin B
4. build plugin B only
5. in an attempt to publish B only, a development version of A gets published as well

If this is a valid case but less troublesome, we can may address it with documentation and/or a printout.

Another question, but not for this PR, should the `artifacts/plugin_bundle.tar.gz` only contain the plugins that are part of the manifest?  Is there value in bundling all versions of all plugins if they are not in the manifest?

File: cmd/plugin/builder/command/cli_compile.go [hunk 1]
```diff
@@ -546,12 +546,22 @@ func getPluginTarget(pd *rtplugin.PluginDescriptor, pluginPath, id string) (stri
...
-	if err != nil {
-		return err
-	}
+	log.Info("saving plugin manifest...")
 
 	if groupByOSArch {
+		existingManifest, err := helpers.ReadPluginManifest(filepath.Join(artifactsDir, cli.PluginManifestFileName))
+		if err == nil {
+			// Merge existing manifest with the newly generated manifest file if no error
+			// Note that we are appending existing manifest to generated one
```
Comment (marckhouzam) [inline cmd/plugin/builder/command/cli_compile.go:555]:
```suggestion
			// Note that we are appending the existing manifest to the generated one
```

File: cmd/plugin/builder/command/cli_compile.go [hunk 2]
```diff
@@ -546,12 +546,22 @@ func getPluginTarget(pd *rtplugin.PluginDescriptor, pluginPath, id string) (stri
...
+			// which gives generated one priority if there is existing plugin with
```
Comment (marckhouzam) [inline cmd/plugin/builder/command/cli_compile.go:556]:
```suggestion
			// which gives the generated one priority if there is already an existing plugin with
```

File: cmd/plugin/builder/command/cli_compile.go [hunk 3]
```diff
@@ -546,12 +546,22 @@ func getPluginTarget(pd *rtplugin.PluginDescriptor, pluginPath, id string) (stri
...
+			// different version
```
Comment (marckhouzam) [inline cmd/plugin/builder/command/cli_compile.go:557]:
```suggestion
			// a different version.
```

File: cmd/plugin/builder/command/cli_compile.go [hunk 4]
```diff
@@ -571,6 +585,29 @@ func savePluginManifest(manifest cli.Manifest, artifactsDir string, groupByOSArc
 	return nil
 }
 
+// mergePluginManifest merges manifest2 to manifest1 file giving manifest1
```
Comment (marckhouzam) [inline cmd/plugin/builder/command/cli_compile.go:588]:
```suggestion
// mergePluginManifest merges 'incomingManifest' into 'baseManifest' giving 'baseManifest'
```

**Comment by anujc25** [comment]
> Thanks for doing this. I'm trying to understand the developer workflow here.
> 
> The flow you are trying to address I assume is:
> 
> 1. in a repo with plugins A and B
> 2. build plugin A only
> 3. then build plugin B only
> 4. then publish both plugins together
> 
> But what if the flow is:
> 
> 1. in a repo with plugins A and B
> 2. build plugin A only
> 3. do development for a couple of weeks on plugin B
> 4. build plugin B only
> 5. in an attempt to publish B only, a development version of A gets published as well
> 
> If this is a valid case but less troublesome, we can address it with documentation and/or a printout.

Thanks for the detailed example. Yes, I feel this scenario is more of a corner case as this should not be an issue for CI systems which are used mostly for publishing plugins. But as you mentioned I also feel that adding documentation clearly explaining this is important. So, I have updated `docs/plugindev/README.md` with the additional details explaining this.

**Review by marckhouzam — APPROVED** [review]
Nice doc!  Thanks

LGTM

**Review by chandrareddyp — APPROVED** [review]
LGTM!

File: cmd/plugin/builder/command/cli_compile.go [hunk 5]
```diff
@@ -571,6 +587,29 @@ func savePluginManifest(manifest cli.Manifest, artifactsDir string, groupByOSArc
...
+	for _, p := range incomingManifest.Plugins {
+		// if plugin not found in base manifest append
+		if findpluginInManifest(baseManifest, p) == nil {
+			mergedManifest.Plugins = append(mergedManifest.Plugins, p)
+		}
+	}
+	return mergedManifest
+}
+
+// findpluginInManifest checks if the plugin already specified in the manifest or not
```
Comment (vuil) [inline cmd/plugin/builder/command/cli_compile.go:603]: nit: findPluginInManifest

plugin already -> "plugin is already"

maybe just take a note to update in the followup.

**Pull request merged**